### PR TITLE
[Polymetis] Fix ABI incompatibility between Pinocchio and PyTorch

### PR DESCRIPTION
**Pull request by 1heart**
# Description

Finally works!

Polymetis has historically required a custom PyTorch version, compiled with CXX11 ABI; this is because the default PyTorch packages are compiled with the old ABI / compilers for widespread compatibility per PEP guidelines, while the default Pinocchio packages are compiled with the newer ABI. This pull request factorizes Polymetis into a Pinocchio-only library, a Torch-only library, and the rest of the code. It hides symbols between them while exposing a C-only(-ish) interface for each library, with the result that these incompatibilities are resolved. This allows us to compile using whichever version the user has installed at the moment, and (in future PRs) unpinning us to custom versions of PyTorch and cleaning up our (previously very specific) dependency tree.

Changes:

- Create `pinocchio_isolation`, which wraps Pinocchio-dependent code in a C interface
- Create `torch_isolation`, which wraps Torch-dependent in a C interface and links to `pinocchio_isolation`
- Refactor Polymetis to link to the above libraries
  - Seems like there's some strange compilation error with Eigen 3.4, so pinning to <3.4

## Type of change

Please check the options that are relevant.

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] Proposes a change (non-breaking change that isn't necessarily a bug)
- [x] Refactor
- [ ] New feature (non-breaking change that adds a new functionality)
- [ ] Breaking change (fix or feature that would break some existing functionality downstream)
- [ ] This is a unit test
- [ ] Documentation only change
- [ ] Datasets Release
- [ ] Models Release

## Type of requested review

- [x] I want a thorough review of the implementation.
- [ ] I want a high level review. 
- [ ] I want a deep design review.

## Before and After

You can compile with the latest version of PyTorch 1.10 in your conda environment and ignoring the custom CXX11 ABI version from the [fair-robotics anaconda channel](https://anaconda.org/fair-robotics/pytorch).

# Testing

Please describe the tests that you ran to verify your changes and how we can reproduce.

# Checklist:

- [ ] I have performed manual end-to-end testing of the feature in my environment.
- [ ] I have added Docstrings and comments to the code.
- [ ] I have made changes to existing documentation where needed.
- [ ] I have added tests that show that the PR is functional.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have added relevant collaborators to review the PR before merge.

### CODE REVIEW

**Comment by 1heart** [comment]
@soumith I isolated the Torch includes to only files within `torchscript_operators`.

Next steps from what I understand:
1. They're still using the C++ standard library, though. What needs to change here -- rewrite it in pure C, or limit usage of the standard library? (Can you provide an example within PyTorch that follows this?)
1. What do I need to do with the current PyTorch extensions (torchrot, and especially torchscript_pinocchio)?
1. Compile as a library separately using the official PyTorch binaries (cxx11_abi=0)
1. Link to it from Polymetis CMakeList

File: polymetis/polymetis/include/polymetis/polymetis_server.hpp
Comment (exhaustin) [inline polymetis/polymetis/include/polymetis/polymetis_server.hpp:62]: shared_ptr?

**Review by exhaustin — CHANGES_REQUESTED** [review]
Thanks for making this happen. Tbh I don't know what most of the isolation changes are doing, but looks reasonable on a high level. I also like the isolation of controller-related operations into an object.

A lot of the real time stack is being changed, so we definitely need to test extensively on hardware before merging. I can run the hardware CI tests.

File: polymetis/polymetis/torch_isolation/include/torch_server_ops.hpp
Comment (1heart) [reply polymetis/polymetis/torch_isolation/include/torch_server_ops.hpp:32]: I actually tried using smart pointers, but they apparently [don't play well with hidden implementations](https://stackoverflow.com/a/8915248/9654319). I'll implement a destructor.

File: polymetis/polymetis/include/polymetis/polymetis_server.hpp
Comment (1heart) [reply polymetis/polymetis/include/polymetis/polymetis_server.hpp:62]: Same as above, will implement a destructor

File: polymetis/polymetis/src/polymetis_server.cpp
Comment (exhaustin) [reply polymetis/polymetis/src/polymetis_server.cpp:237]: On second thought, maybe it's more straightforward if I stack a PR on top of this one that fixes this issue.

File: polymetis/polymetis/src/polymetis_server.cpp
Comment (1heart) [reply polymetis/polymetis/src/polymetis_server.cpp:277]: This should have the previous behavior, so I don't think so.

Comment (exhaustin) [reply polymetis/polymetis/src/polymetis_server.cpp:277]: Yeah after thinking about it we're not doing anything to make TorchScriptedController thread-safe, so this should be fine.

**Review by exhaustin — APPROVED** [review]
Looks good. Tasks before merging:
- Merging the stacked PR #785 to fix the blocking issue in SetController
- Fixing the valid context test that's not passing
- Testing on hardware, preferrably for long hours

**Comment by 1heart** [comment]
@exhaustin if you could run the hardware CI for a couple of hours when you have time and then just merge it if it looks good

**Comment by exhaustin** [comment]
Ran hardware CI for 1 hour straight. I got less than 10 communication constraint violations, which was definitely an improvement on the stability of SetController. 
Should be good to merge once tests pass.

**Pull request merged**
